File: 5-Problem-A-Day/106. Construct Binary Tree from Inorder and Postorder Traversal.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    int postIdx;
    Map<Integer, Integer> inMap;

    private TreeNode build(int[] postorder, int inLeft, int inRight) {
        if (inLeft > inRight)
            return null;
        TreeNode root = new TreeNode(postorder[postIdx]);
        int inIdx = inMap.get(postorder[postIdx]);
        postIdx -= 1;
        root.right = build(postorder, inIdx + 1, inRight);
        root.left = build(postorder, inLeft, inIdx - 1);
        return root;
    }

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        postIdx = postorder.length - 1;
        inMap = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            inMap.put(inorder[i], i);
        }
        return build(postorder, 0, inorder.length - 1);
    }
}

File: 2023/106. Construct Binary Tree from Inorder and Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int, int> inMap;
        for (int i = 0; i < inorder.size(); i++) {
            inMap[inorder[i]] = i;
        }
        return build(inMap, inorder, postorder, 0, (int)inorder.size() - 1, 0, (int)postorder.size() - 1);
    }

    TreeNode* build(unordered_map<int, int>& inMap, vector<int>& inorder, vector<int>& postorder, int inStart, int inEnd, int postStart, int postEnd) {
        if (postStart > postEnd) {
            return NULL;
        }
        TreeNode* root = new TreeNode(postorder[postEnd]);
        int inIndex = inMap[postorder[postEnd]];
        root->left = build(inMap, inorder, postorder, inStart, inIndex - 1, postStart, postEnd - (inEnd - inIndex) - 1);
        root->right = build(inMap, inorder, postorder, inIndex + 1, inEnd, postEnd - (inEnd - inIndex), postEnd - 1);
        return root;
    }
};